spouse/spouse_demo.py

```diff
@@ -0,0 +1,552 @@
+
+@labeling_function(resources=dict(other=other))
+def lf_other_relationship(x, other):
+    return NEGATIVE if len(other.intersection(set(x.between_tokens))) > 0 else ABSTAIN
+
+
+# %% [markdown]
+# #### Apply Labeling Functions to the Data
+# We create a list of labeling functions and apply them to the data
+
+# %%
+lfs = [
+    lf_husband_wife,
+    lf_husband_wife_left_window,
+    lf_same_last_name,
+    lf_married,
+    lf_familial_relationship,
+    lf_family_left_window,
+    lf_other_relationship,
+]
+applier = PandasLFApplier(lfs)
+L = applier.apply(dev_df)
+
+
+# %% [markdown]
+# ### Labeling Function Metrics
+#
+# We can use the lf_summary function to measure various coverage related metrics for LFs. If we have gold labeled data, we can also evaluate accuracy.
+#
+# #### Polarity
+# The set of label values the LF can output when it doesn't abstain. It is common for each LF to have a single polarity.
+#
+# #### Coverage
+# The fraction of candidates that is labeled by our LF.
+#
+# #### Overlaps
+# The fraction of examples labeled by the LF that is also labeled by another LF.
+#
+# #### Conflicts
+# The fraction of examples labeled by the LF that is given a different (non-abstain) label by another LF.
+#
+# #### Correct
+# The number of correctly labeled examples on the gold labeled data.
+#
+# #### Incorrect
+# The number of incorrectly labeled examples on the gold labeled data.
+#
+# #### Empirical Accuracy
+# The fraction of correctly labeled examples on the gold data.
+
+# %%
+from snorkel.labeling.analysis import LFAnalysis
+
+lf_names = [lf.name for lf in lfs]
+LFAnalysis(L).lf_summary(dev_labels, lf_names=lf_names)
+
+# %% [markdown]
+# ## B. Distant Supervision Labeling Functions
+#
+# In addition to using factories that encode pattern matching heuristics, we can also write labeling functions that _distantly supervise_ examples. Here, we'll load in a list of known spouse pairs and check to see if the pair of persons in a candidate matches one of these.
+#
+# **DBpedia**
+# http://wiki.dbpedia.org/
+# Our database of known spouses comes from DBpedia, which is a community-driven resource similar to Wikipedia but for curating structured data. We'll use a preprocessed snapshot as our knowledge base for all labeling function development.
+#
+# We can look at some of the example entries from DBPedia and use them in a simple distant supervision labeling function.
+#
+# Make sure `dbpedia.pkl` is in the `spouse/data` directory.
+
+# %%
+with open("data/dbpedia.pkl", "rb") as f:
+    known_spouses = pickle.load(f)
+
+list(known_spouses)[0:5]
+
+
+# %%
+@labeling_function(
+    resources=dict(known_spouses=known_spouses), preprocessors=[get_person_text]
+)
+def lf_distant_supervision(x, known_spouses):
+    p1, p2 = x.person_names
+    if (p1, p2) in known_spouses or (p2, p1) in known_spouses:
+        return POSITIVE
+    else:
+        return ABSTAIN
+
+
+# %%
+# Helper function to get last name for dbpedia entries.
+def last_name(s):
```

Comment:
maybe this should be hidden in utils.py? it might be confusing to see a function that looks like a preprocessor but isn't being used like one (since it operates over the resource not the data)

or maybe adding a note to point out the difference